crud/src/main/java/com/redhat/lightblue/assoc/ep/ExecutionPlan.java

```diff
@@ -0,0 +1,474 @@
+     */
+    public static Set<Path> getIncludedFieldsOfEntityForProjection(ExecutionBlock block,
+                                                                   CompositeMetadata root,
+                                                                   Projection requestProjection) {
+        Set<Path> fields=new HashSet<>();
+        // What is the path prefix for this entity?
+        CompositeMetadata md=block.getMetadata();
+        Path entityPath=md.getEntityPath();
+        Path globalPrefix=entityPath.isEmpty()?Path.EMPTY:new Path(entityPath,Path.ANYPATH);
+        // entityPath is either empty, meaning this is the root, or a path to a reference field
+        // If entityPath is not empty, then find the projection in the reference field
+        ResolvedReferenceField reference=block.getReference();
+        Projection localProjection;
+        if(reference!=null) {
+            localProjection=reference.getReferenceField().getProjection();
+        } else {
+            localProjection=null;
+        }
+        FieldCursor cursor=md.getFieldCursor();
+        Path skipPrefix=null;
+        Path globalField;
+        while(cursor.next()) {
+            Path localField=cursor.getCurrentPath();
+            globalField=globalPrefix.isEmpty()?localField:new Path(globalPrefix,localField);
+            FieldTreeNode node=cursor.getCurrentNode();
+            if(skipPrefix!=null) {
+                if(!localField.matchingDescendant(skipPrefix)) {
+                    skipPrefix=null;
+                }
+            }
+            if(skipPrefix==null) {
+                if(node instanceof ResolvedReferenceField||
+                   node instanceof ReferenceField) {
+                    skipPrefix=localField;
+                } else  {
+                    if( (node instanceof ObjectField) ||
+                        (node instanceof ArrayField && ((ArrayField)node).getElement() instanceof ObjectArrayElement) ||
+                        (node instanceof ArrayElement)) {
+                        // include its member fields
+                    } else {
+                        if(localProjection!=null&&localProjection.isFieldRequiredToEvaluateProjection(localField)) {
+                            LOGGER.debug("{}: required",localField);
+                            fields.add(localField);
+                        }
+                        if(requestProjection!=null&&requestProjection.isFieldRequiredToEvaluateProjection(globalField)) {
+                            LOGGER.debug("{}: required",localField);
+                            fields.add(localField);
+                        }
+                    }
+                }
+            }
+        }
+        return fields;
+    }
+
+    public static Set<Path> getIncludedFieldsOfRootEntityForSort(CompositeMetadata root,
+                                                                 Sort rootSort) {
+        Set<Path> ret=new HashSet<Path>();
```

Comment:
Don't necessarily need the type info in the right hand assignment anymore.  Only mention it because you use both ways throughout.